### ISSUE TITLE
Merging PE reads for variable-length targets (eg, ITS1)

### ISSUE THREAD

**Issue by werdnus** [benjjneb/dada2#322]
Hi Ben,

I work with fungal metabarcode datasets, and have been really loving being able to use DADA2 since y'all incorporated variable length reads. Thanks for that! But, I've discovered something of a problem with the way that `mergePairs()` handles variable length sequences, particularly those reads that are short enough to sequence all the way through the opposite primer. 

The conversation about whether or not it's worth it to merge paired-end reads is bigger than this issue; suffice it to say, I think that for ITS amplicon data, there are a lot of good reasons to want to assemble paired-end data instead of just using the forward (R1) reads.

After some conversations with other folks about ways to merge pairs, I read [the paper on PEAR](https://academic.oup.com/bioinformatics/article/30/5/614/247231/PEAR-a-fast-and-accurate-Illumina-Paired-End-reAd) and decided I'd like to try using it to assemble my reads.  

But, I wasn't sure how pre-assembling reads would affect the `dada()` denoising, so I decided to pull out a single sample and test differences between using DADA2 with just the R1 reads, with the `mergePairs()` option, and with reads pre-assembled by PEAR. 

![Read Length Distributions](https://user-images.githubusercontent.com/9436670/29989673-503ef606-8f29-11e7-9325-8ed331a2ed88.png)

Here's the read length distributions for the test sample I'm using: if we just use the R1 reads, nothing gets truncated for quality (using the default DADA2 settings), so they're all the same length, whatever the truncation length was set to (200bp); if we use DADA2 to merge pairs, we get only about half as many RSVs out, and they're mostly a little bit longer than the R1s, with some reads stretching out to be much longer; with PEAR, the reads again center around the length of the R1 reads, but then the distribution is pretty normal looking, including tails of longer *and shorter* reads.

I was really curious about those shorter reads: what the hell is going on there? So I did some detective work. I took an example of a shorter reads and BLASTED it, then pulled down the sequence that it matched to (99% identity, 93% coverage) and the _S. cereviciae_ (brewer's yeast, AY2358111) sequence to do some alignments. Then, I found that short sequence in the R1 sequence table and pulled out that sequence, to see what the extra length was all about (PEAR seq = 131bp; R1 seq = 200bp).

![Full Alignment](https://user-images.githubusercontent.com/9436670/29989751-e153dbac-8f29-11e7-8e71-1956657c026b.png)

Here's the alignment... I've marked the priming sites, the ITS regions (freely evolving introns; we sequence the ITS1 region using the ITS1F/ITS2 primer pair), and the rRNA genes. You can see that the taxon this sequence represents (a _Gelidatrema_ sp., LC222848) *actually has a really short ITS1 region!*

At the bottom is the read from the R1 set — notice how it continues out to the reverse primer _and beyond_. At the top, you can see the green bars representing how aligned the sequences are... see how it gets messy after the priming site?

![ITS2 priming site alignment close-up](https://user-images.githubusercontent.com/9436670/29989786-1ee93520-8f2a-11e7-8a92-3ea41056b67c.png)

Here's a close up of the reverse priming site: there's just JUNK after the primer in the R1 read. Which the DADA2 quality filtering let through the pipeline, and which makes this a unique RSV as compared to others _with the same ITS1 sequence_.

So, the length variation in the PEAR assembly seems to be real, and there's probably fewer reads in the PEAR assembled data than the R1 data because it's actually helping us to remove junky sequence at the end of the reads. DADA2 doesn't seem to know how to deal with fully overlapping (to the point of overhanging) sequences when merging pairs.

I guess I would suggest there being some way to find the reverse primers in the R1 reads, and use that as a truncation setting in the `filterAndTrim()` function. Might also be nice to see the PEAR algorithm incorporated into the DADA2 package (since it actually is doing a way better job of merging pairs than anything else I've tried), but that might be beyond y'all's scope. 

Let me know if you want them, and I can post the single sample data-files I used for this, and my R script and PEAR settings.

**Comment by werdnus** [comment]
Hmm, interesting follow up note: in further exploring using PEAR to merge reads before de-noising with `dada()`, I was testing with a sample from a different sequencing run, one that had significantly crappier R2 reads. This is the first time I've run into the requirement that there be no Ns in the dereplicated sequences fed to `dada()`. 

Filtering the R2 reads for `maxN=0` throws away about a third of the reads in the sample. If I merge pairs with PEAR and then filter out the sequences with Ns, I only loose about 3% of sequences, because often the Ns are in the overlap region, and can be resolved during merging. The loss of R1 reads resulting from throwing out reads with an R2 containing ambiguous calls results in the number of RSVs based on just the R1s being reduced from 765 to 616. After pairing with `mergePairs()`, there are only 127 RSVs left. In contrast, if I pair with PEAR and then throw out sequences with any Ns before piping it through `dada()`, I'm left with 729 RSVs — pretty close to the 'just the R1' number. 

I think I understand why the divisive partition would be really difficult if ambiguous bases were allowed, but wouldn't it make a lot more sense to throw away sequences with Ns after merging, rather than during the initial filtering? Or am I missing something here?

**Comment by benjjneb** [comment]
A few points on this. First, note this text from [the tutorial](https://benjjneb.github.io/dada2/tutorial.html) where we introduce the idea of turning off truncation for some ITS amplicon strategies:

> If using this workflow on your own data: For common ITS amplicon strategies, it is undesirable to truncate reads to a fixed length due to the large amount of length variation at that locus. That is OK, just leave out truncLen. **Make sure you removed the forward and reverse primers from both the forward and reverse reads though!**

The "make sure" part is meant to inform people of exactly what you discovered (good detective work btw) -- when reads go through, and even past, the other primer, they will interfere with the denoising algorithm, as the distal primer and the nonsense bases after it will be interpreted as real variation.

Our recommended solution to this issue is to use an external program that can remove primers from both the start and the end of reads (if the read reaches the other primer) as well as any bases after the distal primer. In [our FAQ](https://benjjneb.github.io/dada2/faq.html#how-can-i-remove-primersadaptersetc-from-my-amplicon-reads) we mention trimmomatic and cutadapt, but I imagine there are others.

We have not implemented a new full-featured primer removal solution in dada2 because we don't want to reinvent a wheel that others have already done a good job on.

> I think I understand why the divisive partition would be really difficult if ambiguous bases were allowed, but wouldn't it make a lot more sense to throw away sequences with Ns after merging, rather than during the initial filtering? Or am I missing something here?

Because dada2's workflow merged after F and R reads are independently denoised, Ns would have to be carried through the denoising part, and this just isn't worth it for technical and speed reasons.

Long term, we are considering implementing a "merge-first" workflow as well, but that is not imminent.

**ps:** Would you mind sharing one of your samples that you described in the first post? I have a postdoc that is looking into ITS-specific issues, and that could be a helpful example.

**Comment by werdnus** [comment]
Ah, I missed that line in the tutorial! Too wrapped up in my own little set of issues, I suppose. Glad to know it's just me being oblivious, though. I've not used trimmomatic or cutadapt: I'll have to give them a try! I think the FastX toolkit might have a tool for this, too, now that I'm thinking about it. I support a merge-first workflow!

**Comment by seedmicrobes** [comment]
I would also like to merge reads using PEAR, before learning the error rates, because I am able to retain many more reads this way. How would the current dada2 workflow need to change if merging first?  For example, would pre-merging of sequences break core assumptions of the error learning algorithm, or would it still be valid? I know that PEAR adjusts the quality scores in the merged region, depending on matches/mismatches between merged bases and their relative quality scores, which could be problematic.

**Comment by benjjneb** [comment]
We are getting feedback from several different people that a merge-first workflow might be very useful for ITS data in particular. This is great to hear, as my experience has almost entirely been with 16S/18S data which does not have the length variation of ITS. We are actively looking into improving the ITS paired-end workflow, and @nagasridhar will be helping with this.

Right now, we haven't tested using dada2 on PEAR-merged ITS reads, so if you take that approach, it would be best to sanity check your results at the end to make sure things look reasonable.

**Comment by werdnus** [comment]
Awesome, thanks @benjjneb! I'm certainly sanity checking everything as I go, comparing to the R1-only set. Will let you know if anything particularly wonky shows up. If there's anything I can do to help @nagasridhar with their efforts to improve the ITS workflow, don't hesitate to contact me. Might also want to have a look at some this bioRxiv piece on Illumina/ITS issues from @danchurch: http://www.biorxiv.org/content/early/2017/09/10/184960

**Comment by nagasridhar** [comment]
@werdnus Thank you for referring me to the Illumina/ITS issues paper. 
 
I am running the test ITS sample files through DADA2. Kindly let me know if the adapters in those sequences have been trimmed. If they are not, could you share the adapter sequences for those samples. 

Thanks!

**Comment by werdnus** [comment]
@nagasridhar: Yes, the sequencing facility trims the adaptors from the left-hand side (but some sequence through the other primer on the right-hand side, remember). 

For reference, here are some potentially useful sequences for that sample. 

Prime1: `AATGATACGGCGACCACCGAGATCTACAC`-`GG`-`CTTGGTCATTTAGAGGAAGTAA`
Primer2 form: `5' Illumina adaptor`-`linker`-`ITS1F PCR primer`

Primer2: `CAAGCAGAAGACGGCATACGAGAT`-`ACCGGTATGTAC`-`CG`-`GCTGCGTTCTTCATCGATGC`
Primer2 form: `RC of 3' Illumina adaptor`-`barcode`-`linker`-`ITS2 PCR primer seq`
